From abf3587fb67b7f41a338e024eb3cba745b46504f Mon Sep 17 00:00:00 2001
From: Pantelis Antoniou <panto@antoniou-consulting.com>
Date: Mon, 4 Mar 2013 21:46:13 +0200
Subject: [PATCH 108/177] of-overlay: Handle I2C devices already registered by
 new busses.

I2C is special, as in the busses register the child i2c clients.
So do not register the client device twice.

Signed-off-by: Pantelis Antoniou <panto@antoniou-consulting.com>
---
 drivers/of/overlay.c | 29 ++++++++++++++++++-----------
 1 file changed, 18 insertions(+), 11 deletions(-)

diff --git a/drivers/of/overlay.c b/drivers/of/overlay.c
index 41b883f..29078c7 100644
--- a/drivers/of/overlay.c
+++ b/drivers/of/overlay.c
@@ -424,18 +424,25 @@ static int of_overlay_device_entry_change(struct of_overlay_info *ovinfo,
 				de->pdev = pdev;
 
 		} else {
-			pr_debug("%s: creating new i2c_client device "
-					"new_node='%s' %p\n",
-					__func__, de->np->full_name, de->np);
 
-			client = of_i2c_register_device(adap, de->np);
-
-			if (client == NULL) {
-				pr_warn("%s: Failed to create i2c client device "
-						"for '%s'\n",
-						__func__, de->np->full_name);
-			} else
-				de->client = client;
+			client = of_find_i2c_device_by_node(de->np);
+			if (client != NULL) {
+				/* bus already created the device; do nothing */
+				put_device(&client->dev);
+			} else {
+				pr_debug("%s: creating new i2c_client device "
+						"new_node='%s' %p\n",
+						__func__, de->np->full_name, de->np);
+
+				client = of_i2c_register_device(adap, de->np);
+
+				if (client == NULL) {
+					pr_warn("%s: Failed to create i2c client device "
+							"for '%s'\n",
+							__func__, de->np->full_name);
+				} else
+					de->client = client;
+			}
 		}
 
 	} else {
-- 
1.8.1.4

